src/AI/EventMessage.hpp

```diff
@@ -0,0 +1,648 @@
+
+#include <RTTI/RTTIUtil.hpp>
+#include <BsCorePrerequisites.h>
+#include <Math/BsVector3.h>
+#include <scripting/ScriptTypes.hpp>
+
+namespace REGoth
+{
+  namespace AI
+  {
+    struct EventMessage;
+    using SharedEMessage = bs::SPtr<EventMessage>;
+
+    class Waypoint;
+    using HWaypoint = bs::GameObjectHandle<Waypoint>;
+
+    class Character;
+    using HCharacter = bs::GameObjectHandle<Character>;
+
+    class Item;
+    using HItem = bs::GameObjectHandle<Item>;
+
+    enum class EventMessageType
+    {
+      Event,
+      Npc,
+      Damage,
+      Weapon,
+      Movement,
+      Attack,
+      UseItem,
+      State,
+      Manipulate,
+      Conversation,
+      Magic,
+      Mob,
+
+      SndPlay,
+      SndStarted,
+      SndStopp
+    };
+
+    /**
+     * Basic event-message. Contains type.
+     */
+    struct EventMessage : public bs::IReflectable
+    {
+      using SharedEMessage = std::shared_ptr<EventMessage>;
+
+      EventMessage()
+      {
+        subType        = 0;
+        messageType    = EventMessageType::Event;
+        isJob          = false;
+        deleted        = false;
+        isHighPriority = false;
+        inUse          = false;
+        isOverlay      = false;
+      }
+
+      virtual ~EventMessage()
+      {
+      }
+
+      /**
+       * Type of class this can be casted to
+       */
+      EventMessageType messageType;
+
+      /**
+       * Object this was targeted at
+       */
+      bs::HSceneObject targetObject;
+
+      /**
+       * Object this came from
+       */
+      bs::HSceneObject sourceObject;
+
+      /**
+       * Message subtype defined by the child-classes
+       */
+      unsigned int subType;
+
+      /**
+       * Whether this should wait for other messages to complete before triggering
+       */
+      bool isJob;
+
+      /**
+       * Flag used for cleanup
+       */
+      bool deleted;
+
+      /**
+       * Whether this should overwrite existing messages
+       */
+      bool isHighPriority;
+
+      /**
+       * Whether this is currently playing
+       */
+      bool inUse;
+
+      /**
+       * Whether this message should not block further execution of the message-queue
+       */
+      bool isOverlay;
+
+      /**
+       * Whether this is the first time this message is being handled
+       */
+      bool isFirstRun;
+
+      /**
+       * External callbacks to trigger if this message gets processed.
+       */
+      bs::Event<void(SharedEMessage)> onMessageDone;
+
+      REGOTH_DECLARE_RTTI_FOR_REFLECTABLE(EventMessage)
+    };
+
+    struct NpcMessage : public EventMessage
+    {
+      NpcMessage()
+      {
+        messageType = EventMessageType::Npc;
+        isJob       = true;
+      }
+
+      virtual ~NpcMessage()
+      {
+      }
+
+      REGOTH_DECLARE_RTTI_FOR_REFLECTABLE(NpcMessage)
+    };
+
+    struct DamageMessage : public NpcMessage
+    // FIXME: Reimplement that using a sound handle
+    {
+      enum class DamageSubType
```

Comment:
May want to follow the naming conventions here aswell like `ST_Once` etc. and also make it a plain `enum`, not `enum class`, otherwise casting to `unsigned int` will be a bit cumbersome. One could also explicitely derive the enum from `unsigned int` for it to be even more clear
```cpp
enum DamageSubType : unsigned int
{
  ST_Once,
  ST_PerFrame,
  ST_Max
};
```
Would have to do it for the others aswell.